### DISCUSSION TITLE
Real vs. complex arithmetic - modeling acoustic wave equation in frequency domain

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#19259] [Q&A General] [answered]
Hi
I have the following differential equation which is the acoustic wave equation in the frequency domain, so the primary variable U , the forcing function, and the boundary conditions are complex functions in space. 
Omega, c are the frequency and the acoustic wave velocity.
![image](https://user-images.githubusercontent.com/88537645/139483661-5162a9b5-90e3-49c3-a7bb-87caba8362b6.png)
My question is, given that most, if not all, of the kernels in moose support only the real arithmetic, how is it difficult to implement the former equation within moose? is it easier to go with real arithmetics and have an alternative formulation for the problem or we should stick to the complex arithmetic and implement our own kernels?

Any suggestions?

**Comment by GiudGiud** [comment, marked as answer]
@somu15 @cbolisetti any idea on this?

**Reply by cbolisetti** [reply]
Unfortunately, I don't know. I had this very question myself when I was thinking about the implementation of an absorbing boundary called perfectly matched layer. Perhaps @dschwen or @fdkong has some ideas?

**Reply by fdkong** [reply]
PETSc support complex arithmetic operations very well, and it is easy to get that setup. The challenging part is in libmesh/moose since we do not really have many cases needing complex operations. 

I think that a possible way would be to separate the system into two parts: Real and Image, and then use multiapp or other systems to couple them together. You literally do two "real" solves.

@cticenhour How did you do for the Navier-Stokes-Maxwell system?

**Reply by aaelmeli** [reply]
> Unfortunately, I don't know. I had this very question myself when I was thinking about the implementation of an absorbing boundary called perfectly matched layer. Perhaps @dschwen or @fdkong has some ideas?

That is very similar t what I am intending to do, implementing absorber layers of PML type.
 How far have you reached w.r.t this? or it was just a question that came to your mind?; I am curious.

**Reply by cbolisetti** [reply]
I was looking into one specific formulation for PML and just wanted to see how much effort that would be. I think it might be quite a bit of effort mainly since we don't have anything like it in MOOSE currently. So I didn't really proceed from there.
